sklearn/tree/_splitter.pyx

```diff
@@ -229,6 +242,31 @@ cdef class Splitter:
 
         return self.criterion.node_impurity()
 
+    cdef inline bint check_monotonicity(self, INT32_t monotonic_cst, double lower_bound, double upper_bound):
```

Comment:
This can be marked `nogil`: 
```suggestion
    cdef inline bint check_monotonicity(
        self,
        INT32_t monotonic_cst,
        double lower_bound,
        double upper_bound,
    ) nogil:
```

allowing not taking the GIL when calling it.